pvlib/pvsystem.py

```diff
@@ -1794,73 +1859,165 @@ def singlediode(photocurrent, saturation_current, resistance_series,
-    out['i_x'] = i_x
-    out['i_xx'] = i_xx
+    else:
+        # Calculate points on the IV curve using either 'newton' or 'brentq'
+        # methods. Voltages are determined by first solving the single diode
+        # equation for the diode voltage V_d then backing out voltage
+        v_from_i_fun = partial(singlediode_methods.bishop88_v_from_i,
+                               method=method.lower())
+        i_from_v_fun = partial(singlediode_methods.bishop88_i_from_v,
+                               method=method.lower())
+        mpp_fun = partial(singlediode_methods.bishop88_mpp,
+                          method=method.lower())
+        args = (photocurrent, saturation_current, resistance_series,
+                resistance_shunt, nNsVth)  # collect args
+        v_oc = v_from_i_fun(0.0, *args)
+        i_mp, v_mp, p_mp = mpp_fun(*args)
+        out = OrderedDict()
+        out['i_sc'] = i_from_v_fun(0.0, *args)
+        out['v_oc'] = v_oc
+        out['i_mp'] = i_mp
+        out['v_mp'] = v_mp
+        out['p_mp'] = p_mp
+        out['i_x'] = i_from_v_fun(v_oc / 2.0, *args)
+        out['i_xx'] = i_from_v_fun((v_oc + v_mp) / 2.0, *args)
+        # calculate the IV curve if requested using bishop88
+        if ivcurve_pnts:
+            vd = v_oc * (
+                    (11.0 - np.logspace(np.log10(11.0), 0.0,
+                                        ivcurve_pnts)) / 10.0
+            )
+            i, v, p = bishop88(vd, *args)
+            out['i'] = i
+            out['v'] = v
+            out['p'] = p
+    return out
 
-    # create ivcurve
-    if ivcurve_pnts:
-        ivcurve_v = (np.asarray(v_oc)[..., np.newaxis] *
-                     np.linspace(0, 1, ivcurve_pnts))
 
-        ivcurve_i = i_from_v(resistance_shunt, resistance_series, nNsVth,
-                             ivcurve_v.T, saturation_current, photocurrent).T
+def mpp(photocurrent, saturation_current, resistance_series, resistance_shunt,
+        nNsVth, method='brentq'):
+    """
+    Given the single diode equation coefficients, calculates the maximum power
+    point (MPP).
 
-        out['v'] = ivcurve_v
-        out['i'] = ivcurve_i
+    Parameters
+    ----------
+    photocurrent : numeric
+        photo-generated current [A]
+    saturation_current : numeric
+        diode reverse saturation current [A]
+    resistance_series : numeric
+        series resitance [ohms]
+    resistance_shunt : numeric
+        shunt resitance [ohms]
+    nNsVth : numeric
+        product of thermal voltage ``Vth`` [V], diode ideality factor ``n``, and
+        number of serices cells ``Ns``
+    method : str
+        either ``'newton'`` or ``'brentq'``
 
-    if isinstance(photocurrent, pd.Series) and not ivcurve_pnts:
-        out = pd.DataFrame(out, index=photocurrent.index)
+    Returns
+    -------
+    OrderedDict or pandas.Datafrane
+        ``(i_mp, v_mp, p_mp)``
 
+    Notes
+    -----
+    This function is an option to :func:`singlediode`. Use it when you just
```

Comment:
consider: Use this function when you only want to find the maximum power point. Use singlediode when you need to find additional points on the IV curve.

Reply:
✔️